tortoise/README.md

```diff
@@ -0,0 +1,167 @@
+- hare output
+    
+    tortoise uses output from the hare consensus for voting on blocks within hdist.
+
+- validity/decisions
+
+    for layer before hdist the vote is casted according to the decision. 
+
+- local threshold
+
+    in case when decision wasn't reached but margin from counted weight crosses local threshold tortoise will cast a vote according to the local threshold.
+
+- weak coin
+
+    otherwise tortoise votes according to the coin recorded in the last layer. if sufficient number of honest votes were casted this way then tortoise will either reach decision or atleast vote according to the local threshold in future layers. if honest nodes disagreed on the coin value then the voting will continue until they agree.
+
+## Full tortoise
+
+Full tortoise counts votes from every ballot on every block before the ballot. Once counted weight cross global threshold tortoise can make a decision on a block. Layer is finalized only when all blocks have a decision.
+
+| ballot/block | weight | 0x11 | 0x22 | 0x33 | 0x44 |
+| ------------ | ------ | ---- | ---- | ---- | ---- |
+| 0xaa         | 10     | 1    | -1   | -1   | 1    |
+| 0xbb         | 20     | 1    | 1    | -1   | 1    |
+| total        | -      | 30   | 10   | -30  | 30   |
+| threshold    | -      | 20   | 20   | 20   | 20   |
+| decision     | -      | 1    | 0    | -1   | 1    |
+
+In the above example there are two ballots (0xaa and 0xbb). They vote for 4 blocks. Blocks 0x11 and 0x44 are valid and can be applied to the ledger, block 0x33 is invalid, and block 0x22 is still undecided.
+
+#### Votes from ballots with wrong beacon
+
+Because a smesher can select arbitrary beacon, there is an attack that will allow malicious smeshers to concentrate weight in a specific layer. If tortoise counts malicious weight immediatly it will make erroneous decisions (decide that a block is valid when in fact it is not). In order to prevent such attack tortoise will delay counting ballots with a wrong beacon until enough weight from honest smeshers was counted. 
+
+Note that we can't discard ballots with wrong beacon completely as recovery from partition requires counting ballots from both sides. And they will likely have different beacons if it was a long partition, or occured at the time when tortoise beacon was computed.
+
+#### Scaling issues
+
+Full tortoise complexity grows with the number of layers that are not finalized. 
+
+During live-tortoise this can be a problem only in extreme cases when recent layer fails to be finalized. Due to the excessive malicious voting (> 2/3 fraction of total weight) or insufficient voting from honest smeshers (due to the network problems).
+
+However, during so called rerun (or sync) tortoise needs to count all votes. Otherwise there is always a risk that votes from latest layers changed decision that was made without counting them. With optimized full tortoise it might be practical to count votes when there are 2000 undecided layers, but not higher.
```

Comment:
what does "count all votes" mean here? do you mean during rerun/sync we need to run full mode?

Reply:
in this case i meant count votes from all layers, not necessarily all votes from all layers, that depends on how we count. will try to rephrase